smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelParser.java

```diff
@@ -763,15 +691,27 @@ private void addTraits(ShapeId id, List<TraitEntry> traits) {
      * @param isAnnotation Set to true to indicate that the value for the trait was omitted.
      */
     private void onDeferredTrait(ShapeId target, String traitName, Node traitValue, boolean isAnnotation) {
-        onShapeTarget(traitName, traitValue.getSourceLocation(), id -> {
-            if (isAnnotation) {
-                visitor.onAnnotationTrait(target, id, traitValue.expectNullNode());
-            } else {
-                visitor.onTrait(target, id, traitValue);
-            }
+        modelFile.addForwardReference(traitName, (id, typeProvider) -> {
+            modelFile.onTrait(target, id, coerceTraitValue(id, traitValue, isAnnotation, typeProvider));
         });
     }
 
+    private Node coerceTraitValue(ShapeId traitId, Node value, boolean isAnnotation,
+            Function<ShapeId, ShapeType> typeProvider) {
+        if (isAnnotation && value.isNullNode()) {
+            ShapeType targetType = typeProvider.apply(traitId);
+            if (targetType != null) {
+                if (targetType == ShapeType.STRUCTURE || targetType == ShapeType.MAP) {
```

Comment:
Couldn't this also be a union or document?

Reply:
No— a union requires a value as does a document.